src/main/java/week6/ParkingLotTest.java

```diff
@@ -0,0 +1,151 @@
+            + "move from 10 to 7\n"
+            + "move from 4 to 10\n"
+            + "move from 6 to 4\n"
+            + "move from 10 to 6\n";
+    int[] target = IntStream.rangeClosed(0, 10).map(i -> 10 - i).toArray();
+    int[] source = IntStream.rangeClosed(0, 10).toArray();
+    assertEquals(res, ParkingLot.rearrangeCars(source, target).toString());
+  }
+
+  @Test
+  public void testReverseWithEmptyInMiddle() {
+    String res =
+        "move from 4 to 5\n"
+            + "move from 6 to 4\n"
+            + "move from 3 to 6\n"
+            + "move from 7 to 3\n"
+            + "move from 2 to 7\n"
+            + "move from 8 to 2\n"
+            + "move from 1 to 8\n"
+            + "move from 9 to 1\n"
+            + "move from 0 to 9\n"
+            + "move from 10 to 0\n";
+    int[] source = {1, 2, 3, 4, 5, 0, 6, 7, 8, 9, 10};
+    int[] target = IntStream.rangeClosed(0, 10).map(i -> 10 - i).toArray();
+    assertEquals(res, ParkingLot.rearrangeCars(source, target).toString());
+  }
+
+  @Test
+  public void testZeroCorrectlyPlacedBefore() {
+    String res =
+        "move from 1 to 2\n"
+            + "move from 0 to 1\n"
+            + "move from 3 to 0\n"
+            + "move from 1 to 3\n";
+    int[] source = {1, 2, 0, 3};
+    int[] target = {3, 0, 2, 1};
+    assertEquals(res, ParkingLot.rearrangeCars(source, target).toString());
+  }
+
+  @Test
+  public void testZeroCorrectlyPlacedBeforeAndReverse() {
+    String res =
+        "move from 1 to 2\n"
+            + "move from 0 to 1\n"
+            + "move from 3 to 0\n"
+            + "move from 1 to 3\n"
+            + "move from 4 to 1\n"
+            + "move from 7 to 4\n"
+            + "move from 1 to 7\n"
+            + "move from 5 to 1\n"
+            + "move from 6 to 5\n"
+            + "move from 1 to 6\n";
+    int[] source = {1, 2, 0, 3, 5, 6, 7, 8};
+    int[] target = {3, 0, 2, 1, 8, 7, 6, 5};
+    assertEquals(res, ParkingLot.rearrangeCars(source, target).toString());
+  }
+
+  @Test
+  public void testAlreadyCorrect() {
+    int[] arr = {1, 2, 0, 3, 5, 6, 7, 8};
+    assertEquals("", ParkingLot.rearrangeCars(arr, arr).toString());
+  }
+
+  @Test
+  public void testRandomPattern() {
+    String res =
+        "move from 5 to 2\n"
+            + "move from 4 to 5\n"
+            + "move from 9 to 4\n";
+    int[] source = {1, 2, 0, 3, 5, 6, 7, 8, 9, 4};
+    int[] target = {1, 2, 6, 3, 4, 5, 7, 8, 9, 0};
+    assertEquals(res, ParkingLot.rearrangeCars(source, target).toString());
+  }
+
+  @Test
+  public void testSwap() {
+    String res =
+        "move from 0 to 3\n"
+            + "move from 2 to 0\n"
+            + "move from 1 to 2\n"
+            + "move from 3 to 1\n"
+            + "move from 4 to 3\n"
+            + "move from 5 to 4\n"
+            + "move from 3 to 5\n";
+    int[] source = {1, 2, 3, 0, 4, 5};
+    int[] target = {3, 1, 2, 0, 5, 4};
+    assertEquals(res, ParkingLot.rearrangeCars(source, target).toString());
+  }
+}
```

Comment:
Most of these tests are brittle - they test that you produce a particular sequence of moves, not that the moves do what you want. If you change the algorithm's implementation (including things like changing the HashMap so that the iterator gives you a different arbitrary element - if you really care about consistency of order, look into a LinkedHashMap), the moves may change but it will still work - but the tests will fail, causing a lot of hassle for the maintainer.

I suggest writing a checker that takes a source and target array, and a list of moves, and checks that applying the moves does indeed take you from the source array to the target. Such tests won't be brittle.